File: 2016_11_20/mergesort/mergesort.js
#!/usr/bin/nodejs
"use strict";

function merge(left, right) {
	let i = 0;
	let j = 0;
	let m = [];
	while (i < left.length && j < right.length) {
		if (left[i] <= right[j]) {
			m.push(left[i]);
			i++;
		} else {
			m.push(right[j]);
			j++;
		}
	}
	while (i < left.length) {
		m.push(left[i]);
		i++;
	}
	while (j < right.length) {
		m.push(right[j]);
		j++;
	}
	return m;
}

function mergesort(a) {
	if (a.length <= 1)
		return a;
	let mid = Math.floor(a.length / 2);
	let left = mergesort(a.slice(0, mid));
	let right = mergesort(a.slice(mid));
	return merge(left, right);
}

let size = parseInt(process.argv[2]);
let test = [];
for (let i = 0; i < size; i++)
	test.push(i);
for (let i = 0; i < size; i++) {
// Working on a new box and didn't realize I had an older version of nodejs installed. TIL to be careful about begign ES6 features.
	let r = Math.floor(Math.random() * (size - i)) + i;
	let t = test[i];
	test[i] = test[r]
	test[r] = t;
}

console.log(test);
let out = mergesort(test);
console.log(out);

File: 2016_11_20/mergesort/mergesort.lua
#!/usr/bin/lua

function merge(left, right)
	local i = 1
	local j = 1
	local m = {}
	while i <= #left and j <= #right do
		if left[i] <= right[j] then
			table.insert(m, left[i])
			i = i + 1
		else
			table.insert(m, right[j])
			j = j + 1
		end
	end
	while i <= #left do
		table.insert(m, left[i])
		i = i + 1
	end
	while j <= #right do
		table.insert(m, right[j])
		j = j + 1
	end
	return m
end

function mergesort(a)
	if #a <= 1 then return a end
	local mid = math.floor(#a / 2)
	local left = mergesort({table.unpack(a, 1, mid)})
	local right = mergesort({table.unpack(a, mid+1, #a)})
	return merge(left, right)
end

local size = tonumber(arg[1])
local test = {}

for i = 1,size do
	test[i] = i
end

math.randomseed(os.time())
for i = 1,size do
	local r = math.random(i, size)
	test[i], test[r] = test[r], test[i]
end

print(table.concat(test, " "))
local out = mergesort(test)
print(table.concat(out, " "))








